### PR TITLE
Validate CSR in separate commissioning step

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
Validation of the CSR is coupled to NOC generation currently. Separating these steps allows more flexibility for custom CommissioningDelegates. Also adds remaining values required for CSR validation to the GenerateNOCChain command, so the OperationalCertificateIssuer classes can validate the CSR themselves.
Fixes: #11908

#### Change overview
- make Validate CSR a seprate commissioning stage
- adds parameters for CSR validation to the GenerateNOCChain command

#### Testing
Manually commissioned - checked that the ValidateCSR stage was properly entered and returned success.

### CODE REVIEW

File: src/controller/CHIPDeviceController.cpp [hunk 1]
```diff
@@ -1082,19 +1103,15 @@ CHIP_ERROR DeviceCommissioner::ProcessCSR(DeviceProxy * proxy, const ByteSpan &
...
 
     if (mFabricInfo != nullptr)
     {
         mOperationalCredentialsDelegate->SetFabricIdForNextNOCRequest(mFabricInfo->GetFabricId());
     }
 
-    return mOperationalCredentialsDelegate->GenerateNOCChain(NOCSRElements, AttestationSignature, dac, ByteSpan(), ByteSpan(),
-                                                             &mDeviceNOCChainCallback);
+    return mOperationalCredentialsDelegate->GenerateNOCChain(NOCSRElements, csrNonce, AttestationSignature, attestationChallenge,
+                                                             dac, ByteSpan(), ByteSpan(), &mDeviceNOCChainCallback);
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:1114]: Can you pass previously read PAI and here?

File: src/controller/ExampleOperationalCredentialsIssuer.cpp [hunk 2]
```diff
@@ -176,12 +176,17 @@ CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChainAfterValidation(
     return NewNodeOperationalX509Cert(noc_request, pubkey, mIntermediateIssuer, noc);
 }
 
-CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChain(const ByteSpan & csrElements,
-                                                                 const ByteSpan & attestationSignature, const ByteSpan & DAC,
+CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChain(const ByteSpan & csrElements, const ByteSpan & csrNonce,
+                                                                 const ByteSpan & attestationSignature,
+                                                                 const ByteSpan & attestationChallenge, const ByteSpan & DAC,
                                                                  const ByteSpan & PAI, const ByteSpan & PAA,
```
Comment (tcarmelveilleux) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:182]: The PAA will never be available here. We should remove it from the AapI

File: src/controller/ExampleOperationalCredentialsIssuer.cpp [hunk 3]
```diff
@@ -176,12 +176,17 @@ CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChainAfterValidation(
...
                                                                  Callback::Callback<OnNOCChainGeneration> * onCompletion)
 {
     VerifyOrReturnError(mInitialized, CHIP_ERROR_INCORRECT_STATE);
+    // At this point, Credential issues may wish to validate the CSR information
```
Comment (tcarmelveilleux) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:186]: Nit: issuer

**Review by tcarmelveilleux — APPROVED** [review]

File: src/controller/ExampleOperationalCredentialsIssuer.cpp (hunk 2 above)
Comment (msandstedt) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp:182]: And even if it was, it couldn't be consumed by the entity on the other side of this interface for attestation to be meaningful.

**Review by msandstedt — APPROVED** [review]
PAA needs to be removed from the noc-chain-generation interface and PAI should be piped through from the commissioner.

Else, looks good.  Thanks for cleaning this up.

File: src/controller/CHIPDeviceController.cpp [hunk 4]
```diff
@@ -1071,9 +1091,10 @@ CHIP_ERROR DeviceCommissioner::ProcessCSR(DeviceProxy * proxy, const ByteSpan &
     MATTER_TRACE_EVENT_SCOPE("ProcessOpCSR", "DeviceCommissioner");
     VerifyOrReturnError(mState == State::Initialized, CHIP_ERROR_INCORRECT_STATE);
 
-    ChipLogProgress(Controller, "Getting certificate chain for the device from the issuer");
+    // This will be used once the GenerateNOCChain
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1094]: This comment seems to be missing its end.

File: src/controller/ExampleOperationalCredentialsIssuer.cpp (hunk 3 above)
Comment (bzbarsky-apple) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:186]:
```suggestion
    // At this point, Credential issuer may wish to validate the CSR information
```

**Review by bzbarsky-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp (hunk 1 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp]: passed in now in next commit.

File: src/controller/ExampleOperationalCredentialsIssuer.cpp (hunk 2 above)
Comment (cecille) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp]: Removed this in the latest update.

File: src/controller/CHIPDeviceController.cpp (hunk 4 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp]: Stale bit from the first commit - removed now.

**Review by msandstedt — APPROVED** [review]

**Pull request merged**